Detectors/PHOS/simulation/src/Detector.cxx

```diff
@@ -0,0 +1,1017 @@
+  }
+
+  Double_t lostenergy = fMC->Edep();
+  if (lostenergy < DBL_EPSILON && !isNewPartile)
+    return false; // do not create hits with zero energy deposition
+
+  if (!mGeom)
+    mGeom = Geometry::GetInstance();
+
+  //  if(strcmp(mc->CurrentVolName(),"PXTL")!=0) //Non need to check, alwais there...
+  //    return false ; //  We are not inside a PBWO crystal
+
+  Int_t moduleNumber;
+  fMC->CurrentVolOffID(10, moduleNumber); // get the PHOS module number ;
+  Int_t strip;
+  fMC->CurrentVolOffID(3, strip);
+  Int_t cell;
+  fMC->CurrentVolOffID(2, cell);
+  Int_t detID = mGeom->RelToAbsId(moduleNumber, strip, cell);
+
+  if (superParent == mCurentSuperParent && detID == mCurrentCellID && mCurrentHit) {
+    // continue with current hit
+    mCurrentHit->AddEnergyLoss(lostenergy);
+    return true;
+  }
+
+  // try to find existing Hit
+  if (!isNewPartile) {
+    for (Int_t itr = mHits->size() - 1; itr >= 0; itr--) {
+      Hit* h = &(mHits->at(itr));
+      if (h->GetTrackID() != superParent) // switched to another SuperParent, do not search further
+        break;
+      if (h->GetDetectorID() == detID) { // found correct hit
+        h->AddEnergyLoss(lostenergy);
+        mCurentSuperParent = superParent;
+        mCurrentTrackID = partID;
+        mCurrentCellID = detID;
+        mCurrentHit = h;
+        return true;
+      }
+    }
+  }
+  // Create new Hit
+  Double_t posX = 0., posY = 0., posZ = 0., momX = 0, momY = 0., momZ = 0., energy = 0.;
+  fMC->TrackPosition(posX, posY, posZ);
+  fMC->TrackMomentum(momX, momY, momZ, energy);
+  Double_t estart = fMC->Etot();
+  Double_t time = fMC->TrackTime() * 1.e+9; // time in ns?? To be consistent with EMCAL
+
+  mCurrentHit = AddHit(superParent, detID, Point3D<float>(posX, posY, posZ), Vector3D<float>(momX, momY, momZ), estart,
+                       time, lostenergy);
+  mCurentSuperParent = superParent;
+  mCurrentTrackID = partID;
+  mCurrentCellID = detID;
+  return true;
+}
+
+Hit* Detector::AddHit(Int_t trackID, Int_t detID, const Point3D<float>& pos, const Vector3D<float>& mom, Double_t totE,
+                      Double_t time, Double_t eLoss)
+{
+  LOG(DEBUG4) << "Adding hit for track " << trackID << " with position (" << pos.X() << ", " << pos.Y() << ", "
+              << pos.Z() << ") and momentum (" << mom.X() << ", " << mom.Y() << ", " << mom.Z() << ")  with energy "
+              << totE << " loosing " << eLoss << std::endl;
+  mHits->emplace_back(trackID, detID, pos, mom, totE, time, eLoss);
+  return &(mHits->back());
+}
+
+void Detector::ConstructGeometry()
+{
+  // Create geometry description of PHOS depector for Geant simulations.
+
+  using boost::algorithm::contains;
+  LOG(DEBUG) << "Creating PHOS geometry\n";
+
+  phos::GeometryParams* geom = phos::GeometryParams::GetInstance("Run2");
+
+  if (!geom) {
+    LOG(ERROR) << "ConstructGeometry: PHOS Geometry class has not been set up.\n";
+  }
+
+  if(!fMC)
+    fMC = TVirtualMC::GetMC() ;
```

Comment:
do you need this? Is fMC not initialized at this point?

Reply:
I was surprised but if was not set yet.

Reply:
Thanks. This is likely a bug which needs to be fixed. I will take a separate look.